File: c++/level2/땅따먹기/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

int solution(vector<vector<int> > land)
{
    int answer = 0;

    int row = (int)land.size();

    //init
    vector<vector<int> > dp(row, vector<int>(4,0));
    for(int x = 0 ; x < 4 ; x++)
        dp[row-1][x]= land[row-1][x];

    for(int y = row-2 ; y >=0 ; y--)
    {
        for(int x = 0 ; x < 4 ; x++)
        {
            //
            int tmp = dp[y+1][x];
            dp[y+1][x] = 0;

            //
            dp[y][x] = max(land[y][x]+dp[y+1][0], land[y][x]+dp[y+1][1]);
            dp[y][x] = max(dp[y][x], land[y][x]+dp[y+1][2]);
            dp[y][x] = max(dp[y][x], land[y][x]+dp[y+1][3]);

            //
            dp[y+1][x] = tmp;
        }
    }

    for(int x = 0 ; x < 4 ; x++)
        answer = max(answer, dp[0][x]);

    return answer;
}

int main() {

// url: https://programmers.co.kr/learn/courses/18/lessons/1880

    vector<vector<int> > land = {{1,2,3,5},
                                 {5,6,7,8},
                                 {4,3,2,1}};

    printf("%d", solution(land));

    return 0;
}

File: python/level2/땅따먹기/solution.py
# def hopscotch(board, size):
#     for i in range(size):
#         if i == size - 1:
#             return max(board[i])
#         for j in range(4):
#             # print j, board[i][:j] + board[i][j + 1:]
#             board[i + 1][j] += max(board[i][:j] + board[i][j + 1:])


def solution(land):

    dp = [[0, 0, 0, 0] for _ in range(len(land))]

    dp[0] = land[0]

    for y in range(1, len(land)):
        for x in range(0, 4):
            for k in range(0, 4):
                if k != x:
                    dp[y][x] = max(dp[y][x], land[y][x]+dp[y-1][k])

    return max(dp[-1])


if __name__ == "__main__":
    # url: #

    # input
    input_sample = [[1, 2, 3, 5], [5, 6, 7, 8], [4, 3, 2, 1]]
    # solve
    solution(input_sample)
    # print(solution(input_sample))
